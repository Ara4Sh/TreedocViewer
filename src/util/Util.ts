export default class Util {
  static nonBlankStartsWith(str: string, pattern: string[], detectLength= 1000): boolean {
    for (let i = 0; i < detectLength && i < str.length; i++) {
      if (this.isBlank(str[i]))
        continue;
      for (const p of pattern) {
          if (str.startsWith(p, i))
            return true;
        }
      return false;
      }
    return false;
  }

  static nonBlankEndsWith(str: string, pattern: string[], detectLength= 1000): boolean {
    for (let i = str.length - 1; i >= 0 && i > str.length - detectLength; i--) {
      if (this.isBlank(str[i]))
        continue;
      for (const p of pattern) {
        if (str.endsWith(p, i + 1))
          return true;
      }
      return false;
    }
    return false;
  }

  static isBlank(str: string): boolean {
    return ' \n\r\t'.indexOf(str) >= 0;
  }

  /** Read head string in str upto length, and terminate the at the last line break within this head string */
  static topLines(str: string, length: number): {length: number, numLines: number} {
    const result = { length: -1, numLines: 0 };
    for (let i = Math.min(str.length, length); i >= 0; i--) {
      if (str[i] === '\n') {
        if (result.length < 0) 
          result.length = i + 1;
        result.numLines ++;
      }
    }
    return result;
  }

  public static doIf(condition: boolean, action: () => void) { if (condition) action(); }
}
