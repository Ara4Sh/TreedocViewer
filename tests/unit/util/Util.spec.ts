import Util from '../../../src/util/Util';
import { describe, expect, test } from 'vitest'

describe('Util.ts', () => {
  test('nonBlankStartsWith', () => {
    expect(Util.nonBlankStartsWith('   abcdefghijklmn', ['ab', 'bc]'], 5)).toBeTruthy();
    expect(Util.nonBlankStartsWith('   bcdefghijklmn', ['ab', 'bc]'], 5)).toBeFalsy();
  });

  test('nonBlankEndsWith', () => {
    expect(Util.nonBlankEndsWith('   abcdefghijklmn  \t', ['mn', 'n'], 5)).toBeTruthy();
    expect(Util.nonBlankEndsWith('   bcdefghijklm \n\t  ', ['mn', 'n'], 5)).toBeFalsy();
  });

  test('topLines', () => {
    expect(Util.topLines('12\n34\n56\n', 10)).toMatchSnapshot();
    expect(Util.topLines('12\n34\n56\n', 7)).toMatchSnapshot();
  });
});
