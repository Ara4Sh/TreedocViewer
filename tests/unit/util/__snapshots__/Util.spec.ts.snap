// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Util.ts topLines 1`] = `
Object {
  "length": 9,
  "numLines": 3,
}
`;

exports[`Util.ts topLines 2`] = `
Object {
  "length": 6,
  "numLines": 2,
}
`;
